### PR TITLE
PEP 705: TypedMapping

### PR DESCRIPTION
**Pull request by alicederyn**
Proposal to add a TypedMapping protocol to the language.

Sponsor: @pablogsal

### CODE REVIEW

**Comment by pablogsal** [comment]
I confirm I am sponsoring this PEP 👍

**Comment by hugovk** [comment]
Please use PEP number 705.

(Re-allocating from https://github.com/python/peps/pull/2988 which was closed because it didn't yet find a sponsor.)

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:75]:
```suggestion
Attempting to pass a ``MovieRecord`` into ``movie_string`` results in the error (using mypy):

.. code-block:: text
```

* Fix grammar
* Use appropriate lexer for code block, instead of trying to parse it as Python by default

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:80]:
```suggestion
user from passing a ``MovieRecord`` into a ``Movie`` parameter in the general case, because
```

* Really needs a comma here to separate these clauses and avoid a runon
* Isn't the user passing a ``MovieRecord``-annotated _argument_ into a ``Move``-annotated _parameter_ here?

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:82]:
```suggestion
the type constraints in ``MovieRecord`` (e.g. by ``movie["year"] = None`` or ``del movie["year"]``).
```

Further break up this run-on sentence, and clearly delineate that these are examples.

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:102]:
```suggestion
Rationale
======

The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and supports the cases exemplified above.
```

* IMO, this is really your Rationale section here, since it is explaining the benefits for your proposed _solution_, as opposed to laying out the _problem_,.
* Fix grammar (singular/plural agreement)

Also, it would be really nice to avoid overloading "type"/"types" so much to refer to three different things here, but I'm not sure exactly what to suggest.

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:114]:
```suggestion
* The runtime type of a ``TypedMapping`` object is not constrained to be a ``dict``
* No mutator methods (``__setitem__``, ``__delitem__``, ``update``, etc.) will be generated
* A class definition defines a ``TypedMapping`` protocol if and only if ``TypedMapping`` appears directly in its class bases
* Subclasses can narrow field types, in the same manner as other protocols
```

* Use a bulleted rather than ordered list for items that aren't sequential and don't have a required, unambiguous order
* Use consistent and correct capitalization
* Add missing literal to one ``TypedMapping`` instance

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:116]:
```suggestion
All current and future features of ``TypedDict`` are applicable to ``TypedMapping``, including class-based and alternative syntax, totality, and ``Required``/``NotRequired`` from :pep:`655`.
```

* Avoid potentially confusing/difficult to read double "and"
* Be consistent about using literals, at least within the same section

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:118]:
```suggestion
As with :pep:`589`, this PEP provides a sketch of how a type checker is expected to support type checking operations involving ``TypedMapping`` and ``TypedDict`` objects, but details are left to implementors. In particular, type compatibility should be based on structural compatibility.
```

* Be consistent about using literals or not, at least within the same section

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:128]:
```suggestion
* is the structural intersection of its parents, or invalid if no such intersection exists
* instances must be a dict subclass
* adds mutator methods only for fields it explicitly (re)declares
```

I was very confused about the numbering here, as not only is there no unambiguous declared sequence, but it does not appear to be a direct continuation of the previous numbered list, nor was there another obvious reasons for this.

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:136]:
```suggestion
    class MovieRecord(Movie, TypedDict):
```

Do you mean ``Movie`` here, to match with the description above? ``A`` isn't defined anywhere up to this point.

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:143]:
```suggestion
    movie["year"] = 1985  # Fine; mutator methods added in definition
    movie["name"] = "Terminator"  # Type check error; "name" mutator not declared
```

Fix grammar errors (comma splices)

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:167]:
```suggestion
Type consistency rules
```

Either title case or sentence case is fine for non-top-level headings, so long as you pick one and are (relatively) consistent about it. My suggestion changes this one, but you could equally well change the previous two, if you preferred.

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:259]:
```suggestion
* Declaring methods directly on a ``TypedMapping`` class. Methods are a kind of property, but declarations on a ``TypedMapping`` class are defining keys, so mixing the two is potentially confusing. Banning methods also makes it very easy to decide whether a ``TypedDict`` subclass can mix in a protocol or not (yes if it's just TypedMappings, no if there's a Protocol).


Copyright
=========

This document is placed in the public domain or under the
CC0-1.0-Universal license, whichever is more permissive.
```

Add missing required copyright section. Also, fix missing period and trailing blank lines.

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
**Standard reminder: You can directly apply all the suggestions you want in one go with [`Files changed`](https://github.com/python/peps/pull/2997/files) -> <kbd>Add to batch</kbd> -> <kbd>Commit</kbd>**

## Basic requirements (all PEP Types)

* [x] Read and followed [PEP 1](https://peps.python.org/1) & [PEP 12](https://peps.python.org/12)
* [x] File created from the [latest PEP template](https://github.com/python/peps/blob/main/pep-0012/pep-NNNN.rst?plain=1)
* [x] PEP has next available number, & set in filename (``pep-NNNN.rst``), PR title (``PEP 123: <Title of PEP>``) and ``PEP`` header *[resolved in suggestion]*
* [x] Title clearly, accurately and concisely describes the content in 79 characters or less
* [x] ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly
* [x] ``PEP-Delegate``, ``Topic``, ``Requires`` and ``Replaces`` headers completed if appropriate *[resolved in suggestion]*
* [x] Core dev/PEP editor listed as author or sponsor, and formally confirmed their approval
* [ ] Required sections included
    * [x] Abstract (first section)
    * [x] Copyright (last section; exact wording from template required) *[resolved in suggestion]*
* [x] Code is well-formatted (PEP 7/PEP 8) and is in [code blocks, with the right lexer names](https://peps.python.org/pep-0012/#literal-blocks) if non-Python *[resolved in suggestion]*
* [x] PEP builds with no warnings, pre-commit checks pass and content displays as intended in the rendered HTML
* [x] Authors/sponsor added to ``.github/CODEOWNERS`` for the PEP


## Standards Track requirements

* [x] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) with general agreement that a PEP is appropriate
* [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
    * [x] Motivation
    * [x] Rationale *[resolved in suggestion]*
    * [x] Specification
    * [ ] Backwards Compatibility
    * [ ] Security Implications
    * [ ] How to Teach This
    * [ ] Reference Implementation
    * [x] Rejected Ideas
* [x] ``Python-Version`` set to valid (pre-beta) future Python version
* [x] ~~Any project stated in the PEP as supporting/endorsing/benefiting from it confirms such~~ N/A
* [ ] Right before or after initial merging, [PEP discussion thread](https://peps.python.org/pep-0001/#discussing-a-pep) created and linked to in ``Discussions-To`` and ``Post-History``

File: pep-0705.rst
Comment (JelleZijlstra) [inline pep-0705.rst:5]: The discussions-to thread should be a new thread specifically about the PEP. Since this is very much a type system feature, I would strongly encourage you to have a discussion on typing-sig first.

File: pep-0705.rst
Comment (JelleZijlstra) [inline pep-0705.rst:11]: Technically I don't think "pre-discussions" should be in this list, though they do provide useful context.

File: pep-0705.rst
Comment (JelleZijlstra) [inline pep-0705.rst:112]: We should define exactly what `TypedMapping` does support, not just what it doesn't. For example, does it support the `|` method, which is on `dict` but not `collections.abc.Mapping`?

File: pep-0705.rst
Comment (JelleZijlstra) [inline pep-0705.rst:116]: I would remove "and future". We don't know what future features will be added to TypedDict, and some may not apply to immutable mappings.

File: pep-0705.rst
Comment (JelleZijlstra) [inline pep-0705.rst:118]: Are there concrete reasons why we can't provide a full specification?

**Review by JelleZijlstra — COMMENTED** [review]
This looks quite good and makes a convincing case for the feature, but I have a few comments.

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst:116]: This bothered me too, as a non-typing expert.

Perhaps also a mention that any future proposed `TypedDict` features should explicitly state whether they are also supported on `TypedMapping` or not?

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst:11]: Well, as we've shifted the purpose of this field in modern PEPs toward documenting/linking the history of the PEP's discussion as to opposed to just the dates of its posting, it seems well within the spirit of the current guidance to include it (having written the current letter of it), so long as the topic of the discussion is the idea proposed in the PEP (which it is).

And I agree, it's quite useful. It also helps make it clear that the PEP has fulfilled the requirement:

> [x] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) with general agreement that a PEP is appropriate

And it seems what many authors are already doing as of late, which has been helpful to me as a PEP reader and editor.

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst:5]: And just to clarify, you can wait to create this thread until right before or right after this PR is merged—just make sure to update `Discussions-To` and add an entry to `Post-History` accordingly.

**Comment by alicederyn** [comment]
Just to note I'm still addressing the feedback, just tackled the easier stuff first!

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:246]:
```suggestion
* Declaring methods directly on a ``TypedMapping`` class. Methods are a kind of property, but declarations on a ``TypedMapping`` class are defining keys, so mixing the two is potentially confusing. Banning methods also makes it very easy to decide whether a ``TypedDict`` subclass can mix in a protocol or not (yes if it's just ``TypedMapping`` superclasses, no if there's a ``Protocol``).


Copyright
=========
This document is placed in the public domain or under the
CC0-1.0-Universal license, whichever is more permissive.
```

Update suggestion adding required copyright line after other changes.

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:103]:
```suggestion
All other current features of ``TypedDict`` are applicable to ``TypedMapping``, including class-based and alternative syntax, totality, and ``Required``/``NotRequired`` from :pep:`655`.
```

* Make @JelleZijlstra 's [comment](https://github.com/python/peps/pull/2997#discussion_r1096703542) removing "and future" into a suggestion
* Also explicitly clarify that it is all features except for those listed above

**Review by CAM-Gerlach — COMMENTED** [review]
***Standard reminder: You can directly apply all the suggestions you want in one go with [`Files changed`](https://github.com/python/REPO/pull/NUMBER/files) -> <kbd>Add to batch</kbd> -> <kbd>Commit</kbd>***

Looking good so far! Just a couple quick suggestion followups to the existing fixes, aside from the bigger pending items. Thanks!

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst]: Superseded by [followup suggestion](https://github.com/python/peps/pull/2997#discussion_r1100915336) with just the remaining required copyright change

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst]: Added [a suggestion](https://github.com/python/peps/pull/2997#discussion_r1100916542) for your comment above

**Comment by CAM-Gerlach** [comment]
> Just to note I'm still addressing the feedback, just tackled the easier stuff first!

Thanks @alicederyn !

Just FYI, to help make the easy stuff even easier, in case you aren't aware—instead of recreating the suggestions manually, you can apply them all directly in one go by going to the [`Files changed`](https://github.com/python/peps/pull/2997/files) tab, clicking <kbd>Add to batch</kbd> on the suggestions you want, and then clicking <kbd>Commit</kbd> once you've selected all of them. This also auto-resolves the linked review comments, so they don't need to be double-checked and resolved manually.

**Comment by alicederyn** [comment]
> Just FYI, to help make the easy stuff even easier, in case you aren't aware—instead of recreating the suggestions manually, you can apply them all directly in one go 

Thanks! I was aware of this, but I wanted to check everything rendered correctly locally and iterate, so I did it in the form of a PR to my PR branch. Sorry for creating extra work!

**Comment by alicederyn** [comment]
For reference, I have a bit of a rewrite at https://github.com/alicederyn/peps/pull/4 to address @JelleZijlstra's comments. I'm getting some internal feedback on it before I merge here, but happy for anyone to comment there first.

File: pep-0705.rst
Comment (alicederyn) [reply pep-0705.rst]: Because it's incredibly complex and full of edge cases. For instance, type checkers are still working out how `@final` and `TypedDict` should interact (https://github.com/python/mypy/pull/13838). I think we'd take forever trying to nail a full spec down, and still get it wrong :(

File: pep-0705.rst
Comment (alicederyn) [inline pep-0705.rst:5]: Didn't realise this would block linting so much. Removing until we have a thread to link to.
```suggestion
```

Comment (CAM-Gerlach) [reply pep-0705.rst]: See https://github.com/alicederyn/peps/pull/4#discussion_r1102348723 ; it makes that one check go red as expected, but doesn't interfere with any others.

**Review by CAM-Gerlach — COMMENTED** [review]
LGTM from me, aside from still missing a number of recommended sections (see the checklist above) that would likely be quite helpful to include, even if just briefly for each.

**Comment by alicederyn** [comment]
Backcompat and reference implementation sections at https://github.com/alicederyn/peps/pull/5 for internal review

**Comment by CAM-Gerlach** [comment]
Looks pretty good to me so far, FWIW.

Also, you'll need to fix the expected conflict in the `CODEOWNERS` now that #3009 (PEP 706) was just merged.

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:255]:
```suggestion
``typing_extensions.TypedDict`` in these versions, as the existing ``typing`` implementation
```

* "implementation" is a much more appropriate word than "declaration" here
* Clarify it is the ``typing`` implementation in _existing_ versions but not going forward

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:256]: I would at least briefly explicitly mention why this is a problem.

**Review by CAM-Gerlach — COMMENTED** [review]
Thanks! Just "Security Implications" (which can be a very brief explanation of why there are none, or mentioning how this might improve user security) and "How to Teach This" suggested sections left, assuming there isn't a reason to omit them. Plus, a couple small suggestions/comments.

File: pep-0705.rst
Comment (hugovk) [inline pep-0705.rst:101]: Found via `make spellcheck`:

```suggestion
* By default, all specified keys must be present in a ``TypedMapping`` instance. It is possible to override this by specifying totality, or by using ``NotRequired`` from :pep:`655`.
```

File: pep-0705.rst
Comment (alicederyn) [reply pep-0705.rst:256]:
```suggestion
of ``TypedDict`` does an ``is`` check on the metaclass of all base classes and raises
an error if any other metaclass is found. This would cause any subclass of both
this implementation and a ``TypedMapping`` protocol to trigger an error at runtime.
```
How about this?

Comment (CAM-Gerlach) [reply pep-0705.rst:256]: Looks great, thanks! :+1:

**Comment by alicederyn** [comment]
@CAM-Gerlach I added a "How to teach" section. Pablo suggested leaving out the "security implications" section as it would be empty.

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst]: Seems like this comment was ultimately addressed by replacing this line with a more detailed enumeration of the specific features supported that didn't refer to future features.

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst]: Seems like this was addressed now, @JelleZijlstra ?

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst:11]: Marking this resolved for now as it seems like this is fine and the PEP is otherwise almost ready to merge

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst]: @JelleZijlstra ?

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst]: This was obsoleted and de-facto implemented by further changes

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:261]:
```suggestion
that implementation and a ``TypedMapping`` protocol to trigger an error at runtime.


Security Implications
=====================

There are no known security consequences arising from this PEP.


How to Teach This
```

If there are no security implications, then the general convention is to just explicitly state this for clarity.

File: pep-0705.rst
Comment (CAM-Gerlach) [inline pep-0705.rst:264]: Generally speaking, PEPs are primarily change proposals and relying on them as the main living, user-friendly documentation for a feature should really be avoided. However, in this case this is mostly up to the typing community, which I understand is working on a plan to have a canonical specs and/or docs site like the packaging community does, and can be decided and refined later if and when the PEP is accepted.

Comment (CAM-Gerlach) [inline pep-0705.rst:264]: In any case, link the module too:

```suggestion
This PEP should be linked in the :mod:`typing` module's documentation.
```

**Review by CAM-Gerlach — COMMENTED** [review]
A couple last comments/suggestions from my side; otherwise this LGTM!

File: pep-0705.rst
Comment (alicederyn) [reply pep-0705.rst:261]: Ah, I looked at the recently-accepted https://peps.python.org/pep-0702/ which doesn't have this section.

File: pep-0705.rst
Comment (alicederyn) [reply pep-0705.rst:264]: I used the recently-accepted https://peps.python.org/pep-0692/ as my model here. The typing module documentation does seem to have a lot of links to PEPs. Would definitely be a fan of that changing.

File: pep-0705.rst
Comment (CAM-Gerlach) [reply pep-0705.rst]: Yeah, we've gotten gradually more thorough lately for new PEPs, especially with the new checklist. It's not a hard requirement if there's a specific reason a section clearly isn't relevant, but it doesn't hurt to be explicit.

**Review by CAM-Gerlach — APPROVED** [review]
Thanks for all your hard work on this, @alicederyn ! @JelleZijlstra might have some last comments, but as far as I'm concerned, LGTM!

File: pep-0705.rst
Comment (JelleZijlstra) [reply pep-0705.rst]: Not entirely convinced. Type narrowing is indeed complicated, but that's different from providing a full specification of how type checking itself should work. In any case, this can be part of the conversation after the PEP is submitted.

File: pep-0705.rst
Comment (JelleZijlstra) [inline pep-0705.rst:255]: This is a bit too much implementation detail. We may well be able to work around this restriction with some creative work in typing_extensions. I would suggest taking out everything starting with "Users wanting to inherit". Such limitations can then be explained in the `typing_extensions` documentation.

File: pep-0705.rst
Comment (JelleZijlstra) [inline pep-0705.rst:281]: I would encourage you to submit a typing-extensions implementation as soon as possible.

File: pep-0705.rst
Comment (JelleZijlstra) [inline pep-0705.rst:270]: It's true that we've historically often linked PEPs in the typing docs, but as mentioned above this isn't an ideal state. I'm hopeful that we'll eventually get to rewrite the docs in a way that doesn't require linking to PEPs, so I don't think this PEP should mandate that the typing docs add this specific link.

**Review by JelleZijlstra — COMMENTED** [review]
I have a few more comments but those could also be addressed later. If you would prefer to have the PEP merged now, let me know and I'll do it.

File: pep-0705.rst
Comment (alicederyn) [reply pep-0705.rst:281]: Understood, thanks.

File: pep-0705.rst
Comment (alicederyn) [reply pep-0705.rst]: Changed to no longer be mandatory, but left as a suggestion to mirror https://peps.python.org/pep-0692/#how-to-teach-this &mdash; do let me know if you'd rather leave it out entirely.

**Comment by alicederyn** [comment]
"Request rereview" has apparently changed behaviour and I have no idea how it now works :/ Apologies if I spammed anyone.

**Comment by CAM-Gerlach** [comment]
> "Request rereview" has apparently changed behaviour and I have no idea how it now works :/ Apologies if I spammed anyone.

Would be curious what specifically you're referring to—I've used this feature routinely over the past several years and haven't noticed any significant UI/UX changes, including just now on this PEP. Though, perhaps it had something to do with the fact that we both triggered it right around the same time, with me having loaded the page prior to you pressing the button?

**Comment by alicederyn** [comment]
alicederyn requested review from JelleZijlstra and hugovk and removed request for hugovk and JelleZijlstra

I hit the request review icon for hugovk and Jelle as usual, but the end state was as if I hadn't, and this is the description of what I apparently did. Maybe it work differently if you have different privileges, and I've just never seen it before because I normally do this on repos where I have edit privileges?

**Comment by alicederyn** [comment]
I've definitely seen this behaviour change a week ago in a private GH server, and those usually lag public GH changes by about a half year.

**Comment by CAM-Gerlach** [comment]
Hmm, strange. I've put in a request about it to the others and GitHub.

**Comment by CAM-Gerlach** [comment]
Congratulations @alicederyn , your PEP is now live at https://peps.python.org/pep-0705/ ! Now's the time to open a PEP discussion thread on Typing-SIG and/or the [PEP category of the Python Discourse](https://discuss.python.org/c/peps/19), and link it in the `Discusions-To` and the `Post-History` here. Thanks!

**Comment by alicederyn** [comment]
Thank you @CAM-Gerlach and @JelleZijlstra for your time and feedback!

**Pull request merged**
